src/jvm/main/org/jetbrains/kotlinx/lincheck/ObjectTraverser.kt

```diff
@@ -39,106 +38,72 @@ import kotlin.coroutines.Continuation
-
+    enumerateObjects(obj, objectNumberMap)
     return objectNumberMap
 }
 
 /**
  * Recursively traverses an object to enumerate it and all nested objects.
  *
  * @param obj object to traverse
- * @param processedObjects a set of already processed objects. Required in case of cyclic references.
  * @param objectNumberMap result enumeration map
  */
-private fun enumerateObjects(obj: Any, processedObjects: MutableSet<Any>, objectNumberMap: MutableMap<Any, Int>) {
+private fun enumerateObjects(obj: Any, objectNumberMap: MutableMap<Any, Int>) {
     if (obj is Class<*> || obj is ClassLoader) return
-    if (!processedObjects.add(obj)) return
-    val objectNumber = getObjectNumber(obj.javaClass, obj)
-    objectNumberMap[obj] = objectNumber
-
-    var clazz: Class<*>? = obj.javaClass
-    if (clazz!!.isArray) {
-        if (obj is Array<*>) {
-            obj.forEach { element -> element?.let { enumerateObjects(it, processedObjects, objectNumberMap) } }
-        }
-        return
-    }
-    while (clazz != null) {
-        clazz.declaredFields.filter { f ->
-            !Modifier.isStatic(f.modifiers) &&
-                    !f.isEnumConstant &&
-                    f.name != "serialVersionUID"
-        }.forEach { f ->
-            try {
-                var value: Any? = readField(obj, f)
-
-                if (isAtomic(value)) {
-                    value = value!!.javaClass.getDeclaredMethod("get").invoke(value)
-                }
+    objectNumberMap[obj] = getObjectNumber(obj.javaClass, obj)
 
-                if (value?.javaClass?.canonicalName == "kotlinx.atomicfu.AtomicRef") {
-                    value = readField(value, value.javaClass.getDeclaredField("value"))
-                }
-                if (value?.javaClass?.canonicalName == "kotlinx.atomicfu.AtomicInt") {
-                    value = readField(value, value.javaClass.getDeclaredField("value"))
-                }
-                if (value?.javaClass?.canonicalName == "kotlinx.atomicfu.AtomicLong") {
-                    value = readField(value, value.javaClass.getDeclaredField("value"))
-                }
-                if (value?.javaClass?.canonicalName == "kotlinx.atomicfu.AtomicBoolean") {
-                    value = readField(value, value.javaClass.getDeclaredField("_value"))
-                }
+    traverseObjectHierarchy(obj) { _, f, value ->
+        if (
+            Modifier.isStatic(f.modifiers) ||
+            f.isEnumConstant ||
+            f.name == "serialVersionUID"
+        ) return@traverseObjectHierarchy null
+
+        try {
+            if (value == null || value is Class<*> || value is ClassLoader) return@traverseObjectHierarchy null
+            objectNumberMap[value] = getObjectNumber(value.javaClass, value)
+
+            var jumpValue: Any? = value
+
+            // we jump through most of the atomic classes
+            if (isAtomic(jumpValue)) {
+                jumpValue = jumpValue!!.javaClass.getDeclaredMethod("get").invoke(jumpValue)
```

Comment:
Why don't just put this special handling of atomic types directly into `traverseObjectHierarchy` function?

Reply:
I don't see why we would want that. The utility function is supposed to traverse the object graph (even if the object is an atomic class), but not to perform sophisticated logic for skipping through fields.

Reply:
Atomic objects are a bit special in this regard. 
In many aspects, `AtomicInteger` is very similar to `Integer`, it is essentially just an integer, accesses to which are atomic. Similarly, `AtomicReference` is just a reference to another object.

I don't see in what cases we might want not to skip atomic classes.
Also, probably make sense to skip boxed primitive types (e.g. `Integer`), and some other classes (at least `String`) --- basically any immutable value type.

Reply:
Seems reasonable 👍

Reply:
Hm, now when I use `traverseObjectGraph` in static memory snapshot restoring, I need to track fields of `AtomicInteger`, `AtomicLong`, etc.
So, maybe jump skipping must be a local optimization for `enumerateObject` method and not a common optimization. Then, I have to return `nextObj` from the callbacks `onField` and `onArrayElement`